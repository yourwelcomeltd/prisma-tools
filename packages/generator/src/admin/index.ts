import { writeFileSync } from 'fs';
import { format } from 'prettier';
import { GeneratorOptions, AdminPagesOptions, SchemaObject } from '@paljs/types';
import { createGraphql } from './createGraphql';
import { mergeSchema } from './mergeSchema';
import { ConvertSchemaToObject } from '@paljs/schema';
import { createFile } from './createFile';

const defaultOptions: GeneratorOptions = {
  prismaName: 'prisma',
  output: './src/graphql',
  excludeFields: [],
  excludeModels: [],
  excludeFieldsByModel: {},
  excludeQueriesAndMutations: [],
  excludeQueriesAndMutationsByModel: {},
};

export class UIGenerator {
  schema: SchemaObject;

  constructor(schemaPath: string | string[]) {
    this.schema =
      typeof schemaPath === 'string'
        ? new ConvertSchemaToObject(schemaPath).run()
        : this.mergeSchemas(schemaPath.map((path) => new ConvertSchemaToObject(path).run()));
  }

  mergeSchemas(schemas: SchemaObject[]) {
    const schema: SchemaObject = { models: [], enums: [] };
    schemas.forEach((s) => {
      schema.models.push(...s.models);
      schema.enums.push(...s.enums);
    });
    return schema;
  }

  async buildSettingsSchema(path = 'adminSettings.json', backAsText?: boolean) {
    const newSchema = mergeSchema(this.schema, path);
    const fileContent = await format(`${JSON.stringify(newSchema)}`, {
      singleQuote: true,
      semi: false,
      trailingComma: 'all',
      parser: 'json',
    });

    !backAsText && writeFileSync(path, fileContent);

    return newSchema;
  }

  async generateGraphql(customOptions?: Omit<Partial<GeneratorOptions>, 'nexusSchema'>) {
    const options: GeneratorOptions = {
      ...defaultOptions,
      ...customOptions,
    };
    return await createGraphql(this.schema, options);
  }

  async generateAdminPages(options?: AdminPagesOptions) {
    const content = options?.pageContent ?? page;
    const generatedText: Record<string, string> = {};
    const filterModels = this.schema.models.filter((model) => !options?.models || options?.models.includes(model.name));

    for (const model of filterModels) {
      const fileContent = await format(content.replace(/#{id}/g, model.name), {
        semi: true,
        trailingComma: 'all',
        singleQuote: true,
        printWidth: 120,
        tabWidth: 2,
        parser: 'babel-ts',
      });
      if (options?.backAsText) {
        generatedText[model.name] = fileContent;
      } else {
        createFile(options?.outPut ?? 'src/pages/admin/models/', `${model.name}.tsx`, fileContent);
      }
    }
    return generatedText;
  }
}

const page = `
import React from 'react';
import PrismaTable from 'components/PrismaTable';

const #{id}: React.FC = () => {
  return <PrismaTable model="#{id}" />;
};

export default #{id};
`;
