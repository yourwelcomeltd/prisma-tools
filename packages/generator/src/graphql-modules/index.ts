import { GeneratorOptions, DMMF } from '@paljs/types';
import { writeFileSync, readFileSync, existsSync } from 'fs';
import { createQueriesAndMutations } from './CreateQueriesAndMutations';
import { Generators } from '../Generators';
import { getInputType } from '@paljs/utils';

export class GenerateModules extends Generators {
  generatedText: {
    models: Record<string, { resolvers?: string; typeDefs?: string; module?: string }>;
    inputs: string;
  } = { models: {}, inputs: '' };

  constructor(schemaPath: string, customOptions?: Partial<GeneratorOptions>) {
    super(schemaPath, { output: 'src/app', ...customOptions });
  }

  async run() {
    await this.createModules();
    await this.createInputs();
    await this.createApp();
  }

  private indexPath = this.output('application.ts');
  private index = existsSync(this.indexPath) ? readFileSync(this.indexPath, { encoding: 'utf-8' }) : defaultAppContent;
  private appModules: string[] = getApplication(this.index);

  async createModules() {
    const models = await this.models();
    const datamodel = await this.datamodel();
    for (const model of models) {
      const dataModel = this.dataModel(datamodel.models, model.name);
      let extendsTypes = '';

      if (!this.appModules.includes(model.name + 'Module')) {
        this.appModules.push(model.name + 'Module');
        this.index = `import { ${model.name}Module } from './${model.name}/${model.name}.module';${this.index}`;
      }
      const modelDocs = this.filterDocs(dataModel?.documentation);
      let fileContent = `${modelDocs ? `"""${modelDocs}"""\n` : ''}type ${model.name} {`;

      model.fields.forEach((field) => {
        if (!this.excludeFields(model.name).includes(field.name)) {
          const dataField = this.dataField(field.name, dataModel);
          const fieldDocs = this.filterDocs(dataField?.documentation);
          if (this.shouldOmit(fieldDocs)) {
            return;
          }
          if (dataField?.kind === 'object' && model.name !== dataField.type) {
            if (!extendsTypes.includes(`extend type ${dataField.type}`)) {
              extendsTypes += `extend type ${dataField.type} {`;
              models
                .find((item) => item.name === dataField.type)
                ?.fields.filter((item) => item.outputType.type === model.name)
                .forEach((item) => {
                  extendsTypes = this.getField(item, extendsTypes, fieldDocs);
                });

              extendsTypes += `}\n\n`;
            }
          } else {
            fileContent = this.getField(field, fileContent, fieldDocs);
          }
        }
      });
      fileContent += `\n}\n\n`;

      fileContent += extendsTypes;
      await this.createFiles(model.name, fileContent);
    }
  }

  async getOperations(model: string) {
    const exclude = this.excludedOperations(model);
    return await createQueriesAndMutations(model, exclude, this);
  }

  private async createFiles(model: string, content: string) {
    const operations = await this.getOperations(model);

    !this.options.backAsText && this.mkdir(this.output(model));

    let resolvers = '';

    if (!this.disableQueries(model)) {
      resolvers += operations.queries.resolver;
      content += operations.queries.type;
    }
    if (!this.disableMutations(model)) {
      resolvers += operations.mutations.resolver;
      content += operations.mutations.type;
    }

    await this.createResolver(resolvers, model);

    await this.createTypes(content, model);
    if (this.options.backAsText) {
      this.generatedText.models[model].module = await this.formation(getModule(model), 'babel-ts');
    } else {
      writeFileSync(this.output(model, `${model}.module.ts`), await this.formation(getModule(model), 'babel-ts'));
    }
  }

  private async createTypes(content: string, model: string) {
    content = `import { gql } from 'graphql-modules';

      export default gql\`
      ${await this.formation(content, 'graphql')}
      \`;
      `;
    if (this.options.backAsText) {
      this.generatedText.models[model].typeDefs = await this.formation(content, 'babel-ts');
    } else {
      writeFileSync(this.output(model, 'typeDefs.ts'), await this.formation(content, 'babel-ts'));
    }
  }

  private async createResolver(resolvers: string, model: string) {
    if (resolvers) {
      resolvers = `import { PrismaProvider } from '../Prisma.provider';
      
      export default {
        ${resolvers}
      }
        `;
      if (this.options.backAsText) {
        this.generatedText.models[model] = { resolvers: await this.formation(resolvers, 'babel-ts') };
      } else {
        writeFileSync(this.output(model, 'resolvers.ts'), await this.formation(resolvers, 'babel-ts'));
      }
    }
  }

  async createInputs() {
    let content = await this.generateSDLInputsString();

    content = `import { gql } from 'graphql-modules';\n
    export default gql\`\n${content}\n\`;\n`;

    if (this.options.backAsText) {
      this.generatedText.inputs = content;
    } else {
      writeFileSync(this.output('inputs', this.withExtension(this.inputName)), await this.formation(content));
    }
  }

  async createApp(): Promise<string | void> {
    const content = await this.formation(App(this.appModules, this.index), 'babel-ts');
    if (this.options.backAsText) {
      return content;
    } else {
      writeFileSync(this.indexPath, content);
    }
  }

  getField(field: DMMF.SchemaField, content: string, docs?: string) {
    content += `
  ${docs ? `"""${docs}"""\n` : ''}${field.name}`;
    if (field.args.length > 0) {
      content += '(';
      field.args.forEach((arg) => {
        const inputType = getInputType(arg, this.options);
        content += `${arg.name}: ${inputType.isList ? `[${inputType.type}]` : inputType.type}
              `;
      });
      content += ')';
    }
    content += `: ${
      field.outputType.isList
        ? `[${field.outputType.type}!]!`
        : `${field.outputType.type}${!field.isNullable ? '!' : ''}`
    }`;
    return content;
  }
}

const getModule = (name: string) => {
  return `import { createModule } from 'graphql-modules';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const ${name}Module = createModule({
  id: '${name}',
  typeDefs,
  resolvers,
});
`;
};

const App = (modules: string[], index: string) => {
  const importObject = index.match(/modules:[\S\s]*?]/);
  if (importObject) {
    const modulesMatch = importObject[0].match(/\[[\S\s]*?]/);
    if (modulesMatch) {
      return index.replace(modulesMatch[0], JSON.stringify(modules).replace(/"/g, ''));
    }
  }
  return '';
};

const getApplication = (text: string) => {
  const importObject = text.match(/modules:[\S\s]*?]/);
  if (importObject) {
    const modules = importObject[0].match(/\[([\S\s]*?)]/);
    if (modules) {
      return modules[1]
        .split(',')
        .filter((a) => !!a.replace(/\s/g, ''))
        .map((a) => a.replace(/\s/g, ''));
    }
  }
  return ['CommonModule'];
};

const defaultAppContent = `import { createApplication } from 'graphql-modules';
import { InputsModule } from './inputs/inputs.module'
import { CommonModule } from './common/common.module';
import { addSelect } from './addSelect';
import { PrismaProvider } from './Prisma.provider';

export const application = createApplication({
  modules: [InputsModule, CommonModule],
  providers: [PrismaProvider],
  middlewares: {
    "*": { "*": [addSelect] }
  },
});`;
