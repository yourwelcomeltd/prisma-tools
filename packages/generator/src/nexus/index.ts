import { DMMF } from '@paljs/types';
import { writeFileSync, existsSync, readFileSync } from 'fs';
import { Generators } from '../Generators';
import { getCrud } from './templates';
import { join } from 'path';
import { getInputType } from '@paljs/utils';

export class GenerateNexus extends Generators {
  generatedText: {
    models: Record<
      string,
      { type?: string; queries: Record<string, string>; mutations: Record<string, string>; index?: string }
    >;
    inputs: string;
    index: string;
  } = { models: {}, index: '', inputs: '' };

  private indexPath = this.output(this.withExtension('index'));
  private indexTS = this.readIndex();
  private indexJS: string[] = [];

  async run() {
    await this.createModels();
    await this.createInputs();
    await this.createIndex();
  }

  async createModels() {
    const models = await this.models();
    const dataModels = await this.datamodel();
    for (const model of models) {
      if (this.isJS) {
        this.indexJS.push(model.name);
      } else {
        const exportString = `export * from './${model.name}'`;
        if (!this.indexTS.includes(exportString)) {
          this.indexTS = `export * from './${model.name}'\n${this.indexTS}`;
        }
      }
      const dataModel = this.dataModel(dataModels.models, model.name);
      const modelDocs = this.filterDocs(dataModel?.documentation);
      const hasList = model.fields.find((f) => f.outputType.isList);
      let fileContent = `${this.getImport(`{ objectType${hasList ? ', list' : ''} }`, 'nexus')}\n\n`;
      fileContent += `${!this.isJS ? 'export ' : ''}const ${model.name} = objectType({
        nonNullDefaults: {
          output: true,
          input: false,
        },
  name: '${model.name}',${modelDocs ? `\ndescription: \`${modelDocs}\`,` : ''}
  definition(t) {
    `;
      model.fields.forEach((field) => {
        if (!this.excludeFields(model.name).includes(field.name)) {
          const dataField = this.dataField(field.name, dataModel);
          const fieldDocs = this.filterDocs(dataField?.documentation);
          const options = this.getOptions(field, fieldDocs);
          if (this.shouldOmit(fieldDocs)) {
            return;
          }
          if (field.outputType.location === 'scalar' && field.outputType.type !== 'DateTime') {
            fileContent += `t${this.getNullOrList(field)}.${field.outputType.type.toLowerCase()}('${field.name}'${
              fieldDocs ? `, {description: \`${fieldDocs}\`}` : ''
            })\n`;
          } else {
            fileContent += `t${this.getNullOrList(field)}.field('${field.name}'${options})\n`;
          }
        }
      });

      fileContent += `},\n})\n\n${this.isJS ? `module.exports = {${model.name}}` : ''}`;
      const path = this.output(model.name);

      if (this.options.backAsText) {
        this.generatedText.models[model.name] = { type: await this.formation(fileContent), queries: {}, mutations: {} };
      } else {
        this.mkdir(path);
        writeFileSync(join(path, this.withExtension('type')), await this.formation(fileContent));
      }

      await this.createIndex(path, ['type'].concat(await this.createQueriesAndMutations(model.name)));
    }
  }

  async createQueriesAndMutations(name: string) {
    const exclude = this.excludedOperations(name);
    const modelIndex: string[] = [];
    if (!this.disableQueries(name)) {
      const queriesIndex: string[] = [];
      const path = this.output(name, 'queries');
      for (const item of this.queries.filter((item) => !exclude.includes(item))) {
        const itemContent = await getCrud(name, 'query', item, this);
        if (this.options.backAsText) {
          this.generatedText.models[name].queries[item] = await this.formation(itemContent);
        } else {
          this.createFileIfNotfound(path, this.withExtension(item), await this.formation(itemContent));
        }
        queriesIndex.push(item);
      }
      if (queriesIndex) {
        modelIndex.push('queries');
        const indexPath = join(path, this.withExtension('index'));
        const indexContent = await this.formation(this.getIndexContent(queriesIndex, indexPath));
        if (this.options.backAsText) {
          this.generatedText.models[name].queries.index = indexContent;
        } else {
          writeFileSync(indexPath, indexContent);
        }
      }
    }

    if (!this.disableMutations(name)) {
      const mutationsIndex: string[] = [];
      const path = this.output(name, 'mutations');
      for (const item of this.mutations.filter((item) => !exclude.includes(item))) {
        const itemContent = await getCrud(name, 'mutation', item, this);
        if (this.options.backAsText) {
          this.generatedText.models[name].mutations[item] = await this.formation(itemContent);
        } else {
          this.createFileIfNotfound(path, this.withExtension(item), await this.formation(itemContent));
        }
        mutationsIndex.push(item);
      }
      if (mutationsIndex) {
        modelIndex.push('mutations');
        const indexPath = join(path, this.withExtension('index'));
        const indexContent = await this.formation(this.getIndexContent(mutationsIndex, indexPath));
        if (this.options.backAsText) {
          this.generatedText.models[name].mutations.index = indexContent;
        } else {
          writeFileSync(indexPath, indexContent);
        }
      }
    }
    return modelIndex;
  }

  async createInputs() {
    const dmmf = await this.dmmf();
    const data = dmmf.schema;
    if (this.isJS) {
      this.indexJS.push(this.inputName);
    } else {
      const exportString = `export * from './${this.inputName}'`;
      if (!this.indexTS.includes(exportString)) {
        this.indexTS = `export * from './${this.inputName}'\n${this.indexTS}`;
      }
    }
    const text: string[] = [`${this.getImport('{ enumType, inputObjectType, objectType }', 'nexus')}`, ''];
    const exportModels: string[] = [];
    if (data) {
      const enums = [...data.enumTypes.prisma];
      if (data.enumTypes.model) enums.push(...data.enumTypes.model);
      enums.forEach((item) => {
        exportModels.push(item.name);
        text.push(`${!this.isJS ? 'export ' : ''} const ${item.name} = enumType({
              name: "${item.name}",
              members: ${JSON.stringify(item.values)},
            })\n`);
      });
      const inputObjectTypes = [...data.inputObjectTypes.prisma];
      if (data.inputObjectTypes.model) inputObjectTypes.push(...data.inputObjectTypes.model);
      inputObjectTypes.forEach((input) => {
        const inputFields =
          typeof this.options?.filterInputs === 'function' ? this.options.filterInputs(input) : input.fields;
        if (inputFields.length > 0) {
          exportModels.push(input.name);
          text.push(`${!this.isJS ? 'export ' : ''} const ${input.name} = inputObjectType({
            nonNullDefaults: {
              input: false,
            },
            name: "${input.name}",
            definition(t) {`);
          inputFields
            .filter((field) => !this.options.excludeInputFields?.includes(field.name))
            .forEach((field) => {
              const inputType = getInputType(field, this.options);

              const fieldConfig: Record<string, string> = {
                type: inputType.type as any,
              };
              const arg = field.isRequired ? '.nonNull' : inputType.isList ? '.list' : '';
              text.push(`t${arg}.field(${JSON.stringify(field.name)}, ${JSON.stringify(fieldConfig)});`);
            });
          text.push(`},\n});\n`);
        }
      });
      const outputObjectTypes: DMMF.OutputType[] = [...data.outputObjectTypes.prisma, ...data.outputObjectTypes.model];
      outputObjectTypes
        .filter((type) => type.name.includes('Aggregate') || type.name.endsWith('OutputType'))
        .forEach((type) => {
          exportModels.push(type.name);
          text.push(`${!this.isJS ? 'export ' : ''} const ${type.name} = objectType({
            nonNullDefaults: {
              output: true,
            },
            name: "${type.name}",
            definition(t) {`);

          type.fields
            .filter((field) => !this.options.excludeInputFields?.includes(field.name))
            .forEach((field) => {
              const fieldConfig: Record<string, string> = {
                type: field.outputType.type as any,
              };
              const arg = field.isNullable ? '.nullable' : field.outputType.isList ? '.list' : '';
              text.push(`t${arg}.field(${JSON.stringify(field.name)}, ${JSON.stringify(fieldConfig)});`);
            });
          text.push(`},\n});\n`);
        });
      if (this.isJS) {
        text.push(`module.exports = {${exportModels.join(',')}}`);
      }
    }
    if (this.options.backAsText) {
      this.generatedText.inputs = await this.formation(text.join('\n'));
    } else {
      writeFileSync(join(this.output(), this.withExtension(this.inputName)), await this.formation(text.join('\n')));
    }
  }

  async createIndex(path?: string, content?: string[]): Promise<string | void> {
    if (path && content) {
      const indexPath = join(path, this.withExtension('index'));
      const fileContent = await this.formation(this.getIndexContent(content, indexPath));
      if (this.options.backAsText) {
        return fileContent;
      } else {
        writeFileSync(indexPath, fileContent);
      }
    } else {
      const fileContent = await this.formation(this.isJS ? this.getIndexContent(this.indexJS) : this.indexTS);
      if (this.options.backAsText) {
        return fileContent;
      } else {
        writeFileSync(this.output(this.withExtension('index')), fileContent);
      }
    }
  }

  private readIndex() {
    return existsSync(this.indexPath) ? readFileSync(this.indexPath, { encoding: 'utf-8' }) : '';
  }

  private getOptions(field: DMMF.SchemaField, docs?: string) {
    const options: any = docs ? { description: docs } : {};
    if (field.outputType.location !== 'scalar' || field.outputType.type === 'DateTime')
      options['type'] = field.outputType.type;
    let toString = JSON.stringify(options);
    if (field.args.length > 0) {
      toString = toString.slice(0, -1);
      toString += ', ';
      toString += this.getNexusArgs(field.args);
    }
    if (field.outputType.location === 'outputObjectTypes') {
      toString = toString.slice(0, -1);
      toString += `, resolve(root${this.isJS ? '' : ': any'}) {
      return root.${field.name}
    },
    }`;
    }
    return ', ' + toString;
  }

  private getNullOrList(field: DMMF.SchemaField) {
    return field.outputType.isList ? '.list' : !field.isNullable ? '' : '.nullable';
  }
}
