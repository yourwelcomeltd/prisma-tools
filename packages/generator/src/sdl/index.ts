import { existsSync, readFileSync, writeFileSync } from 'fs';
import { createQueriesAndMutations } from './CreateQueriesAndMutations';
import { Generators } from '../Generators';
import { GenerateTypes } from './GenerateTypes';
import { getInputType } from '@paljs/utils';

export class GenerateSdl extends Generators {
  generatedText: {
    models: Record<string, { resolvers: string; typeDefs: string }>;
    inputs: string;
  } = { models: {}, inputs: '' };

  async run() {
    await this.createModels();
    await this.createInputs();
    await this.createMaster();
    if (!this.isJS) {
      const generateTypes = new GenerateTypes(await this.dmmf(), this.options);
      const code = await generateTypes.run();
      writeFileSync(this.output('../resolversTypes.ts'), await this.formation(code));
    }
  }

  resolversPath = this.output(this.withExtension('resolvers'));
  resolversIndex = existsSync(this.resolversPath)
    ? readFileSync(this.resolversPath, { encoding: 'utf-8' })
    : defaultResolverFile(this.isJS);
  resolversExport: string[] = getCurrentExport(this.resolversIndex);

  typeDefsPath = this.output(this.withExtension('typeDefs'));
  typeDefsIndex = existsSync(this.typeDefsPath)
    ? readFileSync(this.typeDefsPath, { encoding: 'utf-8' })
    : defaultTypeFile(this.isJS);
  typeDefsExport: string[] = getCurrentExport(this.typeDefsIndex);

  async createModels() {
    const dataModels = await this.datamodel();
    for (const model of await this.models()) {
      const dataModel = this.dataModel(dataModels.models, model.name);
      const modelDocs = this.filterDocs(dataModel?.documentation);
      let fileContent = `${modelDocs ? `"""${modelDocs}"""\n` : ''}type ${model.name} {`;
      const excludeFields = this.excludeFields(model.name);
      model.fields.forEach((field) => {
        if (!excludeFields.includes(field.name)) {
          const dataField = this.dataField(field.name, dataModel);
          const fieldDocs = this.filterDocs(dataField?.documentation);
          if (this.shouldOmit(fieldDocs)) {
            return;
          }
          fileContent += `
          ${fieldDocs ? `"""${fieldDocs}"""\n` : ''}${field.name}`;
          if (field.args.length > 0) {
            fileContent += '(';
            field.args.forEach((arg) => {
              const inputType = getInputType(arg, this.options);
              fileContent += `${arg.name}: ${inputType.isList ? `[${inputType.type}]` : inputType.type}
              `;
            });
            fileContent += ')';
          }
          fileContent += `: ${
            field.outputType.isList
              ? `[${field.outputType.type}!]!`
              : `${field.outputType.type}${!field.isNullable ? '!' : ''}`
          }`;
        }
      });

      fileContent += `\n}\n\n`;
      await this.createFiles(model.name, fileContent);
    }
  }

  private async getOperations(model: string) {
    const exclude = this.excludedOperations(model);
    return await createQueriesAndMutations(model, exclude, this);
  }

  private async createFiles(model: string, typeContent: string) {
    const operations = await this.getOperations(model);
    !this.options.backAsText && this.mkdir(this.output(model));

    let resolvers = '';
    if (!this.disableQueries(model)) {
      resolvers += operations.queries.resolver;
      typeContent += operations.queries.type;
    }
    if (!this.disableMutations(model)) {
      resolvers += operations.mutations.resolver;
      typeContent += operations.mutations.type;
    }
    const resolverFile = await this.createResolvers(resolvers, model);
    const typeFile = await this.createTypes(typeContent, model);
    if (this.options.backAsText) {
      this.generatedText.models[model] = { resolvers: resolverFile || '', typeDefs: typeFile || '' };
    }
  }

  private async createResolvers(resolvers: string, model: string): Promise<string | void> {
    if (resolvers) {
      if (this.isJS) {
        resolvers = `
      const ${model} = {
        ${resolvers}
      }
      
      module.exports = { 
      ${model}
      }
        `;
      } else {
        resolvers = `import { Resolvers } from '../../resolversTypes';
      
      const resolvers: Resolvers = {
        ${resolvers}
      }
      export default resolvers;
        `;
      }

      if (!this.resolversExport.includes(model)) {
        this.resolversExport.push(model);
        this.resolversIndex = `${this.getImport(this.isJS ? `{ ${model} }` : model, `./${model}/resolvers`)}\n${
          this.resolversIndex
        }`;
      }

      if (this.options.backAsText) {
        return this.formation(resolvers);
      } else {
        writeFileSync(this.output(model, this.withExtension('resolvers')), await this.formation(resolvers));
      }
    }
  }

  private async createTypes(fileContent: string, model: string): Promise<string | void> {
    if (this.isJS) {
      fileContent = `const { default: gql } = require('graphql-tag');\n
    const ${model} = gql\`\n${await this.formation(fileContent, 'graphql')}\n\`;\n
    module.exports = { 
      ${model}
      }`;
    } else {
      fileContent = `import gql from 'graphql-tag';\n
    export default gql\`\n${await this.formation(fileContent, 'graphql')}\n\`;\n`;
    }

    if (!this.typeDefsExport.includes(model)) {
      this.typeDefsExport.push(model);
      this.typeDefsIndex = `${this.getImport(this.isJS ? `{ ${model} }` : model, `./${model}/typeDefs`)}\n${
        this.typeDefsIndex
      }`;
    }

    if (this.options.backAsText) {
      return this.formation(fileContent);
    } else {
      writeFileSync(this.output(model, this.withExtension('typeDefs')), await this.formation(fileContent));
    }
  }

  async createInputs() {
    let content = await this.generateSDLInputsString();

    if (this.isJS) {
      content = `const { default: gql } = require('graphql-tag');\n
    const ${this.inputName} = gql\`\n${content}\n\`;\n
    module.exports = { 
      ${this.inputName}
      }`;
    } else {
      content = `import gql from 'graphql-tag';\n
    export default gql\`\n${content}\n\`;\n`;
    }

    if (!this.typeDefsExport.includes(this.inputName)) {
      this.typeDefsExport.push(this.inputName);
      this.typeDefsIndex = `${this.getImport(
        this.isJS ? `{ ${this.inputName} }` : this.inputName,
        `./${this.inputName}`,
      )}\n${this.typeDefsIndex}`;
    }
    if (this.options.backAsText) {
      this.generatedText.inputs = await this.formation(content);
    } else {
      writeFileSync(this.output(this.withExtension(this.inputName)), await this.formation(content));
    }
  }

  async createMaster() {
    writeFileSync(this.resolversPath, await this.formation(replaceExports(this.resolversExport, this.resolversIndex)));

    writeFileSync(this.typeDefsPath, await this.formation(replaceExports(this.typeDefsExport, this.typeDefsIndex)));
  }
}

export const replaceExports = (exports: string[], text: string) => {
  const matchText = text.match(/\[([\S\s]*?)]/);
  if (matchText) {
    return text.replace(matchText[0], JSON.stringify(exports).replace(/"/g, ''));
  }
  return '';
};

const getCurrentExport = (text: string) => {
  const matchText = text.match(/\[([\S\s]*?)]/);
  if (matchText) {
    return matchText[1]
      .split(',')
      .filter((a) => a)
      .map((a) => a.replace(/\s/g, ''));
  }
  return [];
};

const defaultResolverFile = (isJs?: boolean) =>
  isJs
    ? `
    const resolvers = [];
    
    module.exports = {resolvers};`
    : `export default [];`;
const defaultTypeFile = (isJs?: boolean) =>
  isJs
    ? `const { mergeTypeDefs } = require('@graphql-tools/merge');
const { InputTypes } = require('./InputTypes');

const typeDefs = mergeTypeDefs([InputTypes]);

module.exports = {typeDefs};`
    : `import { mergeTypeDefs } from '@graphql-tools/merge';
import InputTypes from './InputTypes';

export default mergeTypeDefs([InputTypes]);`;
