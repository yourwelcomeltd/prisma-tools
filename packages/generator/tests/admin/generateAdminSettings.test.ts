import { join } from 'path';
import { UIGenerator } from '../../src';

describe('Generate Json Object For Admin Tables', () => {
  const schemaPath = join(__dirname, '../schemas/schema.prisma');
  it('Should back with Json Settings', async () => {
    const generator = new UIGenerator(schemaPath);
    expect(await generator.buildSettingsSchema('adminSettings.json', true)).toMatchSnapshot();
  });
});
