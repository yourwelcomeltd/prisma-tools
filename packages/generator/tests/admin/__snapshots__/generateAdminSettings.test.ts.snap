// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Json Object For Admin Tables Should back with Json Settings 1`] = `
{
  "enums": [
    {
      "fields": [
        "ADMIN",
        "USER",
      ],
      "name": "Role",
    },
  ],
  "models": [
    {
      "create": true,
      "delete": true,
      "displayFields": [
        "id",
      ],
      "fields": [
        {
          "create": false,
          "defaultValue": "autoincrement()",
          "editor": false,
          "filter": true,
          "id": "User.id",
          "isId": true,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "id",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Id",
          "type": "Int",
          "unique": false,
          "update": false,
          "upload": false,
        },
        {
          "create": false,
          "defaultValue": "now()",
          "editor": false,
          "filter": true,
          "id": "User.createdAt",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "createdAt",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Created At",
          "type": "DateTime",
          "unique": false,
          "update": false,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "User.email",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "email",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Email",
          "type": "String",
          "unique": true,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "User.name",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "name",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": false,
          "sort": true,
          "title": "Name",
          "type": "String",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "User.password",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "password",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Password",
          "type": "String",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": ""{}"",
          "editor": false,
          "filter": true,
          "id": "User.permissions",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "permissions",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Permissions",
          "type": "Json",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "User.posts",
          "isId": false,
          "kind": "object",
          "list": true,
          "map": undefined,
          "name": "posts",
          "order": -1,
          "read": true,
          "required": false,
          "sort": true,
          "title": "Posts",
          "type": "Post",
          "unique": false,
          "update": true,
          "upload": false,
        },
      ],
      "id": "User",
      "idField": "id",
      "name": "User",
      "update": true,
    },
    {
      "create": true,
      "delete": true,
      "displayFields": [
        "id",
      ],
      "fields": [
        {
          "create": false,
          "defaultValue": "autoincrement()",
          "editor": false,
          "filter": true,
          "id": "Post.id",
          "isId": true,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "id",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Id",
          "type": "Int",
          "unique": false,
          "update": false,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": "false",
          "editor": false,
          "filter": true,
          "id": "Post.published",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "published",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Published",
          "type": "Boolean",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "Post.title",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "title",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Title",
          "type": "String",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": true,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "Post.author",
          "isId": false,
          "kind": "object",
          "list": false,
          "map": undefined,
          "name": "author",
          "order": -1,
          "read": true,
          "required": false,
          "sort": true,
          "title": "Author",
          "type": "User",
          "unique": false,
          "update": true,
          "upload": false,
        },
        {
          "create": false,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "Post.authorId",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "authorId",
          "order": -1,
          "read": true,
          "relationField": true,
          "required": false,
          "sort": true,
          "title": "Author Id",
          "type": "Int",
          "unique": false,
          "update": false,
          "upload": false,
        },
        {
          "create": false,
          "defaultValue": "now()",
          "editor": false,
          "filter": true,
          "id": "Post.createdAt",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "createdAt",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Created At",
          "type": "DateTime",
          "unique": false,
          "update": false,
          "upload": false,
        },
        {
          "create": false,
          "defaultValue": undefined,
          "editor": false,
          "filter": true,
          "id": "Post.updatedAt",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "updatedAt",
          "order": -1,
          "read": true,
          "relationField": false,
          "required": true,
          "sort": true,
          "title": "Updated At",
          "type": "DateTime",
          "unique": false,
          "update": false,
          "upload": false,
        },
      ],
      "id": "Post",
      "idField": "id",
      "name": "Post",
      "update": true,
    },
  ],
}
`;
